### DISCUSSION TITLE
Examples no longer compile on AVR 32u4, AVR 328

### DISCUSSION THREAD

**Discussion by ki4mcw** [jgromes/RadioLib#1270] [Q&A]
Somewhere between version 4 and version 7 of RadioLib, the use of Program Memory expanded to such an extent that the examples no longer compile for AVR 32u4 or 328 (or at least, the examples for RTTY, APRS, and Hellschrieber). Are there any options besides replacing hardware or using old versions of the lib?

<code>
Sketch uses 33536 bytes (116%) of program storage space. Maximum is 28672 bytes.
Global variables use 2503 bytes of dynamic memory.
Sketch too big; see https://support.arduino.cc/hc/en-us/articles/360013825179 for tips on reducing it.
text section exceeds available space in board

Compilation error: text section exceeds available space in board
</code>

Thanks,
Rob KI4MCW

**Comment by jgromes** [comment]
Unfortunately the only two options are exactly what you wrote: use older versions of the library, or change the hardware. I would strongly suggest not using ATmega328/32u4 anymore, both of these designs are 16 years old at this point (at least that's how far back the datasheet revisions go).

With the next major release, I will most likely drop official support for these and remove them from the CI and readme.

Also, this is not really an issue (and there's not much we can do about it), I will convert this thread to a discussion.

**Comment by jrhbcn** [comment]
Hello, I am running into the same problem as this discussion. In my case, I am trying to use RadioLib in an [ATTINY3226](https://www.microchip.com/en-us/product/attiny3226) (with 32 KB Flash, 3 KB SRAM and 256B of EEPROM). ATtinys are a very nice hardware (2-series is also quite recent) and I think they are excellent companions to some LoRa modules. It is quite a pity that RadioLib cannot be compiled in them. 

I have tried to define several `RADIOLIB_EXCLUDE_*` modules and protocols as I only want to use the sx1262 with the LoRa protocol, however, some of them seem to break the compilation (for instance `RADIOLIB_EXCLUDE_Si443x`) and, in any case, they do not seem to affect at all the final compiled size. :(

Is there anything I can do manually to solve this? Such as removing manually some files and #includes or some functionality I would not use?

Many thanks for making public this excellent piece of code.

**Reply by StevenCellist** [reply]
I'm not 100% sure on the class inheritance, but if I'm not mistaken, the radio is instantiated through the PhysicalLayer class, which the compiler isn't able to optimize - it compiles all inherited functions. Without the PhysicalLayer class (or Module class with PhysicalLayer or something like that), the compiler is better able to optimize the compiled binary and you may _just_ be able to fit it. 
I've been able to make RadioLib 6.6's LoRaWAN implementation run with an SX1262 on a ATmega4048, with ~10kB of the resulting binary being the LoRaWAN protocol, by stripping away the Module/PhysicalLayer stuff.

**Reply by HeadBoffin** [reply]
DietLib2, recompiled just now to do a memory refresh, needs 32033 bytes of memory - the OTAA and MacCommands taking up the biggest parts. So as much as I pine for the good old days of a 32K device because the ATtiny range is awesome, it's going to take some serious engineering to make it fit.

**Reply by jrhbcn** [reply]
Thanks for your replies @StevenCellist, @HeadBoffin, I also think keeping support for ATtinys would be awesome,

I see both `OTAA` and `MacCommands` are related to LoRaWAN. Would it be somehow possible (without the serious engineering HeadBoffin refers to :) to remove LoRaWAN functionality (keeping the rest) in order to make it fit? Ideally with a `#define` directive (I already tried with `RADIOLIB_EXCLUDE_LORAWAN`, but size is the same) or maybe commenting out / removing some files directly?

**Reply by StevenCellist** [reply]
Whatever is not in use by your sketch is not compiled in*, so if you don't create a LoRaWANNode variable, the LoRaWAN code isn't in use. That is why none of the excludes has any effect: the compiler is smart enough to not use it.

The problem is in the * and mentioned before: because RadioLib uses multiple derived/inherited classes for the radio module, the compiler has trouble figuring out what is actually in use and what isn't. That is where you look at my earlier comments about Module/PhysicalLayer: strip away those 'translation layers' and the compiler knows what it can remove and what it cannot. So that has nothing to do with LoRaWAN: it's all about the radio instance.

**Reply by StevenCellist** [reply]
(And no: it isn't really (or really isn't) feasible to integrate functionality in the library to have the user choose whether they want to use derived classes or not, unfortunately.)

**Reply by jrhbcn** [reply]
> That is where you look at my earlier comments about Module/PhysicalLayer: strip away those 'translation layers' and the compiler knows what it can remove and what it cannot. 

@StevenCellist, you mean removing the function in PhysicalLayer class and then remove the override in all other module classes?

**Reply by StevenCellist** [reply]
Probably a bit easier to remove their base class inatantiations, such as: 
```
SX126x::SX126x(Module* mod) : PhysicalLayer
```
Removing the `PhysicalLayer` there will automatically remove all its functions but probably also throw warnings about unused overrides so remove the override keywords indeed. Not sure if that change does enough, but it's likely the path you need to go down. It's been a while since I tried it, and not at the desk, so it's all from the top of my head.

**Reply by jrhbcn** [reply]
Success @StevenCellist, thank you for your help! 

I have made SX126X not dependent from PhysicalLayer. I had to remove the overrides and copy some functions/variables over from PhysicalLayer to the SX126X class. Examples now compile, and the size has been reduced to "just" 18786 bytes. 

I don't know if there is any interest in trying to generalize this for other users and add it to the library. Maybe moving the not-overriden functions to a new file (so they can be used in different modules) and then using some defines to use PhysicalLayer as a base or not.

In any case, thanks for your help, I will check if my changes do indeed work and not just compile though... :)

**Reply by jgromes** [reply]
@jrhbcn this is an interesting excercise. If we could get at least the basic drivers (SX126x, SX127x) running on smaller platforms, that might be worth looking into. The downside is that this will make it impossible to use any protocols, which depend on the PHY abstraction. So definitely no LoRaWAN on smaller MCUs.

**Reply by StevenCellist** [reply]
Unless there is a way to forcefully make the compiler behave with pruning of unused overridden functions? No clue.

**Reply by jrhbcn** [reply]
> @jrhbcn this is an interesting excercise. If we could get at least the basic drivers (SX126x, SX127x) running on smaller platforms, that might be worth looking into. The downside is that this will make it impossible to use any protocols, which depend on the PHY abstraction. So definitely no LoRaWAN on smaller MCUs.

That is exactly what I thought, at least having base LoRa functionality in smaller MCUs is great imho. Take a look at my [fork](https://github.com/jrhbcn/RadioLib). It is very rough but I use a define variable RADIOLIB_EXCLUDE_SX126X_PHYSICAL_LAYER to separate sx126x from the PhysicalLayer class. It is still rough as I just copied the non-override functions from the PhysicalLayer to the SX126X class and I defined RADIOLIB_EXCLUDE_DIRECT_RECEIVE also just to take care of less functions. But it could be a start for doing it in a "semi-proper" way.

In any case, thanks again for this excellent library!

**Reply by RobertDeHate** [reply]
I like to use the same library for both sides of my wireless connections because it is easier to ensure compatablity.
I also like to use cheap LoRa32 development boards off Amazon for remote sensors that don't need much resources.
I have wireless sensors in my ductwork that measure pressure and drive servos to control the vents in my home and 32k is plenty of memory. But they need to run lean because they are powered by the wind turbine attached to them in the ventilation system.
So there is a section of the community that would love lightweight libraries.
And it is communities like this that come up with those solutions so I have to thank you guys for the work you do.
Just a use case to consider.
Robert

**Reply by HeadBoffin** [reply]
I'd love to use RL for everything, we put it on a diet and got LW down to a point where it will fit, with user code, in to 48KB of flash. So you should have vanishingly close to zero expectations that it can be made to fit in to older MCUs.

I sort of get what you are saying about power use & size of code, but I suspect that's a very minor power use differential.

No reason why you can't use any of the other very good LoRa P2P libraries at either end - not knowing what you have at the control end it's hard to comment.

**Reply by RobertDeHate** [reply]
Thanks,
Was looking for a more modern implementation. Been using RadioHead.
I have a collection of processors I use,
RP2040/RFM95(8MFlash) ,LoRa32u4ii/SX1276(32kFlash), FeatherM0/RFM95(256kFlash),ASR6502/SX1262(128kFlash),ESP32-S3/SX1262(16MFlash), NRF52840/SX1262(2MFlash)
So having a library that is flexible is needed.
And home automation works great with LoRa radios, I even have a LoRa network running in my RV.
And expendable boards are sometimes needed. Use them once and they get destroyed. So lower cost models have their roll.
But I digress.
I can not get a RH transmitter to talk with a RL receiver.
I am sure it has something to do with the headers not matching.
So I dig deeper into the library to find out.
I see:
`
      \brief Sets LoRa sync word.
      \param syncWord LoRa sync word to be set.
      \param controlBits Undocumented control bits, required for compatibility purposes.`
Have to see how this compares to RH library.
And I ran across something about a header CRC as apposed to the normal CRC.

And this was baffling
`When more bytes than received are requested, only the number of bytes requested will be returned`
SX126x.h line 753   ;-)

Thanks for the good librarys.
At least if I have to hack them they are worth hacking.
Robert

**Reply by jgromes** [reply]
@RobertDeHate 

> I am sure it has something to do with the headers not matching.

If you are referring to the LoRa packet header, then that is fully handled by the LoRa radio. The only thing any control library can do is to completely disable the header (a.k.a. implicit header mode). RadioLib has LoRa header enabled by default, I don't know what RadioHead does but I would be surprised if it was disabled.

The actual things to check are:
* the usual LoRa parameters, SF/BW/CR
* CRC type (polynomial, initial etc.)
* whitening

**Reply by HeadBoffin** [reply]
Are you using different libraries at either end?

If Sandeep's bare bones lib doesn't work for you, then StuartsProjects is by someone that predates pretty much every other LoRa offering and goes outside to test stuff.

**Reply by RobertDeHate** [reply]
I wanted to use the same library on each end, that was why I was talking about 32k flash.
Anyway, I backed off to V6 of RadioLib and it compiles in 32k ;-)
Had issues communicating SX1276 to SX1262.
Checked SyncWord and found, "You couldn't make this up if you tried."
But no changes to SW helped.
It turns out to be the spreading factor, which was defaulted to 6.
I found in the SX1262 data sheet:
`However, these modifications have made the new spreading factor incompatible with previous device generations. Especially, SF6 on the SX1261/2 will not be backward compatible with the SF6 used on the SX1276.`
Changed it to 8, and communication is good now.
Thanks for all.
Robert
